File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<bits/stdc++.h>
using namespace std;

bool c[101][101];
int bfsMap[101][101];
int m, n;
int dx[4] = {0, 1, 0, -1};
int dy[4] = {-1, 0, 1, 0};
queue<pair<int,int>> q;

int solution(vector<vector<int> > maps)
{
    int answer = 0;
    n = maps[0].size();
    m = maps.size();
    
    q.push(make_pair(0,0));
    c[0][0] = true;
    bfsMap[0][0] = 1;
    
    while(!q.empty()){
        int curX = q.front().first;
        int curY = q.front().second;
        
        q.pop();
        
        for(int i = 0; i < 4; i++){
            int nx = curX + dx[i];
            int ny = curY + dy[i];
            
            if(nx < 0 || nx >= m || ny < 0 || ny >= n) continue;
            if(c[nx][ny]) continue;
            if(maps[nx][ny] == 0) continue;
            
            c[nx][ny] = true;
            q.push(make_pair(nx,ny));
            bfsMap[nx][ny] = bfsMap[curX][curY] + 1;
            
        }
    }
    
    if(!bfsMap[m - 1][n - 1]) answer = -1;
      else answer = bfsMap[m - 1][n - 1];
    
    return answer;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n, m = len(maps), len(maps[0])
    dx = [0, 1, 0, -1]
    dy = [-1, 0, 1, 0]
    
    queue = deque([(0, 0)])
    visited = [[False] * m for _ in range(n)]
    visited[0][0] = True
    bfs_map = [[0] * m for _ in range(n)]
    bfs_map[0][0] = 1
    
    while queue:
        x, y = queue.popleft()
        
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            
            if 0 <= nx < n and 0 <= ny < m and not visited[nx][ny] and maps[nx][ny] == 1:
                visited[nx][ny] = True
                bfs_map[nx][ny] = bfs_map[x][y] + 1
                queue.append((nx, ny))
    
    answer = bfs_map[n - 1][m - 1]
    return answer if answer != 0 else -1